src/engine/mod.rs

```diff
@@ -0,0 +1,189 @@
+        unimplemented!()
+    }
+
+    /// Displays a template.
+    ///
+    /// # Failures
+    /// * When the template cannot be found
+    /// * When an error occurred during compilation
+    /// * When an error occurred during rendering
+    pub fn display(&mut self, _path: &str, _data: ()) -> Result<(), TwigError> {
+       unimplemented!()
+    }
+
+    /// Loads and compiles a template.
+    ///
+    /// # Failures
+    /// * When the template cannot be found
+    /// * When an error occurred during compilation
+    pub fn load_template(&mut self, path: &str, _index: Option<u32>) -> Result<template::Compiled, TwigError> {
+        // TODO: Cache compiled templates
+        //  * cache lookup
+        //  * check if cache is fresh
+        //  * store in cache
+
+        let template_raw = try!(self.load_template_raw(path));
+        return self.compile_template(&template_raw);
+    }
+
+    /// Loads raw template.
+    ///
+    /// # Failures
+    /// * When the template cannot be found
+    fn load_template_raw(&mut self, path: &str) -> Result<template::Raw, TwigError> {
+        let loader = try!(self.loader());
+        let source = try_chain!(loader.source(path));
+        Ok(template::Raw::new(source, path))
+    }
+
+    /// Compiles a template.
+    ///
+    /// # Failures
+    /// * When an error occurred during lexing or parsing.
+    fn compile_template(&mut self, template: &template::Raw) -> Result<template::Compiled, TwigError> {
+        let tokenstream = {
+            let lexer = try!(self.lexer());
+            try_chain!(lexer.tokenize(template))
+        };
+
+        let compiled = {
+            let parser = try!(self.parser());
+            try_chain!(parser.parse(&tokenstream))
+        };
+
+        Ok(compiled)
+    }
+
+    /// Sets the engine extensions.
+    pub fn set_extensions(&mut self, ext: ExtensionRegistry) -> &mut Engine {
+        self.ext = Some(Rc::new(ext)); // TODO: switch to callback pattern to provide arguments
+
+        self
+    }
+
+    /// Gets the engine extensions.
+    pub fn extensions(&self) -> Result<&Rc<ExtensionRegistry>, TwigError> {
+        match self.ext {
+            Some(ref ext) => Ok(ext),
+            None => {
+                try_chain!(err!(ExtensionRegistryErrorCode::NotInitialized))
+            }
+        }
+    }
+
+    /// Sets the loader instance.
+    pub fn set_loader(&mut self, loader: Box<Loader>) -> &mut Engine {
+        self.loader = Some(loader); // TODO: switch to callback pattern to provide arguments
+
+        self
+    }
+
+    /// Gets the loader instance.
+    pub fn loader(&mut self) -> Result<&mut Loader, TwigError> {
+        match self.loader {
+            Some(ref mut loader) => return Ok(&mut **loader),
+            None => {
+                return err!(TwigErrorCode::LoaderNotInitialized)
+            }
+        }
+    }
+
+    /// Sets the lexer instance.
+    pub fn set_lexer(&mut self, lexer: Lexer) -> &mut Engine {
```

Comment:
I know the [TwigEnvironment](https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Environment.php) has these setters, but I think they don't make sense in Rust. We should always create them, by default, because they are not polymorphic, and they can't change during the `compile_template`.